lib/std/os/windows.zig

```diff
@@ -153,14 +153,136 @@ pub fn OpenFile(sub_path_w: []const u16, options: OpenFileOptions) OpenError!HAN
     }
 }
 
-pub const CreatePipeError = error{Unexpected};
+pub fn GetCurrentProcess() HANDLE {
+    const process_pseudo_handle: usize = @bitCast(@as(isize, -1));
+    return @ptrFromInt(process_pseudo_handle);
+}
+
+pub fn GetCurrentProcessId() DWORD {
+    return @truncate(@intFromPtr(teb().ClientId.UniqueProcess));
+}
+
+pub fn GetCurrentThread() HANDLE {
+    const process_pseudo_thread: usize = @bitCast(@as(isize, -2));
+    return @ptrFromInt(process_pseudo_thread);
+}
+
+pub fn GetCurrentThreadId() HANDLE {
+    return teb().ClientId.UniqueThread;
+}
+
+pub const CreatePipeError = error{ Unexpected, SystemResources };
 
+/// A Zig wrapper around `NtCreateNamedPipeFile` and `NtCreateFile` syscalls.
+/// It implements similar behavior to `CreatePipe` and is meant to serve
+/// as a direct substitute for that call.
 pub fn CreatePipe(rd: *HANDLE, wr: *HANDLE, sattr: *const SECURITY_ATTRIBUTES) CreatePipeError!void {
-    if (kernel32.CreatePipe(rd, wr, sattr, 0) == 0) {
-        switch (kernel32.GetLastError()) {
-            else => |err| return unexpectedError(err),
-        }
+    // Up to NT 5.2 (Windows XP/Server 2003), `CreatePipe` would generate a pipe similar to:
+    //
+    //      \??\pipe\Win32Pipes.{pid}.{count}
+    //
+    // where `pid` is the process id and count is a incrementing counter.
+    // The implementation was changed after NT 6.0 (Vista) to open a handle to the Named Pipe File System
+    // and use that as the root directory for `NtCreateNamedPipeFile`.
+    // This object is visible under the NPFS but has no filename attached to it.
+    //
+    // This implementation replicates how `CreatePipe` works in modern Windows versions.
+    var name = blk: {
+        const str = std.unicode.utf8ToUtf16LeStringLiteral("\\Device\\NamedPipe\\");
+        const len: u16 = @truncate(str.len * @sizeOf(u16));
+        break :blk UNICODE_STRING{ .Length = len, .MaximumLength = len, .Buffer = @constCast(@ptrCast(str)) };
+    };
+    var attrs = OBJECT_ATTRIBUTES{
+        .ObjectName = @constCast(&name),
+        .Length = @sizeOf(OBJECT_ATTRIBUTES),
+        .RootDirectory = null,
+        .Attributes = 0,
+        .SecurityDescriptor = null,
+        .SecurityQualityOfService = null,
+    };
+
+    var npfs: HANDLE = undefined;
```

Comment:
I think this handle should be cached globally as in the original implementation.

The only slightly tricky part is that you have to do so atomically using cmpxchg and close the local handle based on the return value of that.

Reply:
IMO I don't think it's worth adding some global state for something that (seemingly) isn't called a lot. Open to arguments either way though.

Reply:
If it's something that truly has a global lifetime, then indeed a global variable would be preferable. What would be the purpose of avoiding global state in this case?